File: native/V2StyxLib/lib/server/ClientBalancer.cpp
/*
 * ClientBalancer.cpp
 *
 *  Created on: May 22, 2012
 *      Author: V.Shcriyabets (vshcryabets@gmail.com)
 */

#include "server/ClientBalancer.h"
#include "stdio.h"

ClientBalancer::ClientBalancer(int iounit, IVirtualStyxFile *root, std::string protocol) {
	mAllConnections = new vector<Socket>();
	mHandler = new ClientsHandler(iounit, root, protocol);
}

ClientBalancer::~ClientBalancer() {
	delete mAllConnections;
}

void ClientBalancer::pushNewConnection(Socket socket) {
	printf("New conection\n");
	mNewConnections.push_back(socket);
}

void ClientBalancer::pushReadable(Socket socket) {
//	printf("New data\n");
	mReadable.push_back(socket);
}

void ClientBalancer::process() {

	// new connections
	for (vector<Socket>::iterator socketIterator = mNewConnections.begin();
			socketIterator < mNewConnections.end(); socketIterator++ ) {
		printf("FD=%d\n",*socketIterator);
		mHandler->addClient(*socketIterator);
		mAllConnections->push_back(*socketIterator);
	}
	mNewConnections.clear();

	// new readables
	for (vector<Socket>::iterator socketIterator = mReadable.begin();
			socketIterator != mReadable.end(); socketIterator++ ) {
		bool closed = mHandler->readClient(*socketIterator);
		if ( closed ) {
#ifdef WIN32
			closesocket(*socketIterator);
#else
			::close(*socketIterator);
#endif
			for (vector<Socket>::iterator it = mAllConnections->begin();
					it != mAllConnections->end(); it++ ) {
				if ( *it == *socketIterator ) {
					mAllConnections->erase(it);
					break;
				}
			}
		}
	}
	mReadable.clear();
}
vector<Socket> *ClientBalancer::getAllConnections() {
	return mAllConnections;
}

File: java/library/src/com/v2soft/styxlib/library/server/ClientBalancer.java
package com.v2soft.styxlib.library.server;

import com.v2soft.styxlib.library.core.IMessageProcessor;
import com.v2soft.styxlib.library.messages.base.StyxMessage;
import com.v2soft.styxlib.library.server.vfs.IVirtualStyxFile;

import java.io.IOException;


/**
 *
 * @author V.Shcriyabets (vshcryabets@gmail.com)
 *
 */
public class ClientBalancer implements IMessageProcessor {
    private IMessageProcessor mHandler;

    public ClientBalancer(int iounit, IVirtualStyxFile root, String protocol) {
        mHandler = new TMessagesProcessor(iounit, root, protocol);
    }
    @Override
    public void addClient(ClientState state) {
        mHandler.addClient(state);
    }
    @Override
    public void removeClient(ClientState state) {
        mHandler.removeClient(state);
    }
    @Override
    public void processPacket(StyxMessage message) throws IOException {
        mHandler.processPacket(message);
    }

    @Override
    public int getReceivedPacketsCount() {
        return 0;
    }

    @Override
    public int getReceivedErrorPacketsCount() {
        return 0;
    }

    @Override
    public void close() throws IOException {
        mHandler.close();
    }
}
